### PR TITLE
on_mode_changed callback added

### PR DESCRIPTION
**Pull request by info-rchitect**
First discussed [here](https://stackoverflow.com/questions/48811004/changing-object-data-using-with-mode-method).

~~~ruby
  class MyModel
    include Origen::Model
    attr_reader :my_version
    attr_accessor :mode_changed_count
    def initialize
      @my_version = version
       add_mode :mode1
      @mode_changed_count = 0
    end
    
    def on_mode_changed(options)
      @mode_changed_count += 1
    end
  end

    it 'on_mode_changed callback works' do
      m = MyModel.new
      m.modes.should == [:mode1]
      m.mode.should == nil
      m.mode_changed_count.should == 0
      m.mode = :mode1
      m.mode_changed_count.should == 1
    end
~~~

### CODE REVIEW

File: templates/web/guides/misc/callbacks.md.erb [hunk 1]
```diff
@@ -266,6 +267,15 @@ guaranteed to be called even if the thread crashes.
 An example use cases is to ensure that any temporary files are deleted.
 
 
+<a class="anchor" name="mode_changed"></a>
+
+### Mode Changed
+
+#### on_mode_changed
+
```
Comment (ginty) [inline templates/web/guides/misc/callbacks.md.erb]: Internal links should be created using this helper:

<%= path "api/Origen/Model.html#current_mode=-instance_method" %>

That means they will work when generating offline, and generally if the root url changes, e.g. if we move to https in future.

File: templates/web/guides/misc/callbacks.md.erb [hunk 2]
```diff
@@ -266,6 +267,15 @@ guaranteed to be called even if the thread crashes.
 An example use cases is to ensure that any temporary files are deleted.
 
 
+<a class="anchor" name="mode_changed"></a>
+
+### Mode Changed
+
+#### on_mode_changed
```
Comment (ginty) [inline templates/web/guides/misc/callbacks.md.erb]: Should reflect that this takes an options hash.

Should also note here that the new mode will be passed in via options[:mode].

File: templates/web/guides/misc/callbacks.md.erb (hunk 1 above)
Comment (info-rchitect) [reply templates/web/guides/misc/callbacks.md.erb]: where does the text go?

Comment (ginty) [reply templates/web/guides/misc/callbacks.md.erb]: The above just replaces http://...., so it would be:

`[changes its current mode](<%= ... %>).`

**Comment by info-rchitect** [comment]
thx @ginty.  This will allow us to model the device modes without having to duplicate a bunch of clock information inside the ChipMode object.  The ChipMode object is handy for storing things you haven't or don't want to really model but if an object model exists then this callback seems like a better option.

**Comment by info-rchitect** [comment]
completed [here](https://github.com/Origen-SDK/origen/pull/237)

**Pull request merged**